### PR TITLE
terminalsAndIcons for FMI2 + refactoring

### PR DESCRIPTION
**Pull request by PeterMeisrimelModelon**
Adding current terminalsAndIcons parsing to FMI2 + some refactoring.

Re-factoring:
1. Split XML lists from fmi3_xml_parser.h/c into fmi3_xml_model_description_scheme.h & fmi_xml_terminals_and_icons_scheme.h, merged together again fmi3_xml_parser_scheme.h
2. Split some parsing utility functionality into fmi3_xml_parser_util.h/c
3. Split the implementation of fmi3_xml_parser_context into fmi3_xml_parser_context_impl.h
4. Renamed most terminalsAndIcons stuff (files, structs/types, functions) from fmi3_* to fmi_*, where possible. 
5. Fixed documentation for Terminals and Icons.

For FMI2 + terminalsAndIcons
1. Added parsing of terminalsAndIcons to FMI2
2. Duplicated terminalsAndIcons functions for FMI2 variants, when necessary.
3. Duplicated testutils functionality to FMI2. Could use macros here instead to create the same functions for FMI2 + 3?
4. Duplicated non error specifics tests for terminalsAndIcons for FMI2.

+ various misc cleanup 

**What is not included thus far:**
1. Definite separation of fmi3 & terminalsAndIcons parser, currently using same parser + schema list
2. terminalsAndIcons xml callbacks (vendor annotation parsing). Options are to use 1. Separate callbacks, would require API change or 2. Using the same as FMI2/3, would require re-factoring to unify the FMI2 + 3 ones or 3. another option?

Might be easier to review commit by commit.

### CODE REVIEW

File: src/Import/src/FMI3/fmi3_import.c [hunk 1]
```diff
@@ -102,7 +102,9 @@ fmi3_import_t* fmi3_import_parse_xml(
 
     if (!fmu) {
         context->callbacks->free(xmlPath);
-        if (!terminalsAndIconsPath) context->callbacks->free(terminalsAndIconsPath);
+        if (terminalsAndIconsPath) {
```
Comment (modelonrobinandersson) [inline src/Import/src/FMI3/fmi3_import.c:105]: Would it make sense to wrap this `if (terminalsAndIconsPath) { free ... }` in one function?

File: src/XML/src/FMI/fmi_xml_terminals_and_icons_scheme.h [hunk 2]
```diff
@@ -0,0 +1,62 @@
+/*
+    Copyright (C) 2023 Modelon AB
```
Comment (modelonrobinandersson) [inline src/XML/src/FMI/fmi_xml_terminals_and_icons_scheme.h:2]:
```suggestion
    Copyright (C) 2024 Modelon AB
```
again feel free to ignore.

File: src/XML/src/FMI3/fmi3_xml_vendor_annotations.c [hunk 3]
```diff
@@ -78,30 +78,30 @@ int fmi3_xml_handle_VariableTool(fmi3_xml_parser_context_t *context, const char*
...
-            context->anyToolName = vendor;
-            context->anyParent = 0;
-            context->useAnyHandleFlg = 1;
+        size_t len;
+        fmi3_xml_model_description_t* md = context->modelDescription;
+        jm_vector(char)* bufName = fmi3_xml_reserve_parse_buffer(context,1,100);
+        jm_string *pvendor;
+        char* vendor = 0;
+
+        if(!bufName) return -1;
```
Comment (modelonrobinandersson) [inline src/XML/src/FMI3/fmi3_xml_vendor_annotations.c:87]:
```suggestion
        if (!bufName) return -1;
```

File: src/XML/src/FMI3/fmi3_xml_vendor_annotations.c [hunk 4]
```diff
@@ -78,30 +78,30 @@ int fmi3_xml_handle_VariableTool(fmi3_xml_parser_context_t *context, const char*
...
+        /* <xs:attribute name="name" type="xs:normalizedString" use="required"> */
+        if( fmi3_xml_parse_attr_as_string(context, fmi3_xml_elmID_Tool, fmi_attr_id_name, 1, bufName))
```
Comment (modelonrobinandersson) [inline src/XML/src/FMI3/fmi3_xml_vendor_annotations.c:89]:
```suggestion
        if (fmi3_xml_parse_attr_as_string(context, fmi3_xml_elmID_Tool, fmi_attr_id_name, 1, bufName))
```

File: src/XML/src/FMI3/fmi3_xml_vendor_annotations.c [hunk 5]
```diff
@@ -78,30 +78,30 @@ int fmi3_xml_handle_VariableTool(fmi3_xml_parser_context_t *context, const char*
...
+            return -1;
+        pvendor = jm_vector_push_back(jm_string)(&md->vendorList, vendor);
+        len = jm_vector_get_size(char)(bufName);
+        if(pvendor )
```
Comment (modelonrobinandersson) [inline src/XML/src/FMI3/fmi3_xml_vendor_annotations.c:93]:
```suggestion
        if (pvendor)
```
maybe we should add a `{ }` as well

File: src/XML/src/FMI3/fmi3_xml_vendor_annotations.c [hunk 6]
```diff
@@ -78,30 +78,30 @@ int fmi3_xml_handle_VariableTool(fmi3_xml_parser_context_t *context, const char*
...
+            *pvendor = vendor = (char*)(context->callbacks->malloc(len + 1));
+        if(!pvendor || !vendor) {
```
Comment (modelonrobinandersson) [inline src/XML/src/FMI3/fmi3_xml_vendor_annotations.c:95]:
```suggestion
        if (!pvendor || !vendor) {
```

File: Config.cmake/test_fmi2.cmake [hunk 7]
```diff
@@ -118,7 +113,7 @@ set_target_properties(
 set(FAIL_NAME_CHECK 0)
 set(PASS_NAME_CHECK 1)
 
-add_test(ctest_fmi2_xml_parsing_test fmi2_xml_parsing_test ${FMIL_TEST_DIR}/FMI2/parser_test_xmls/)
+add_test(ctest_fmi2_xml_parsing_test fmi2_xml_parsing_test ${FMI2_TEST_XML_DIR}/)
```
[the remark this answers is not in the document]
Comment (PeterMeisrimelModelon) [reply Config.cmake/test_fmi2.cmake:116]: Actually, yes. And it is still required. 
Copied the comment.

File: src/Import/src/FMI3/fmi3_import.c (hunk 1 above)
Comment (PeterMeisrimelModelon) [reply src/Import/src/FMI3/fmi3_import.c:105]: I don't think so, there is only 2 uses.

There is some code duplication in this function to handle the cleanup in the various possible failure stages.
The corresponding fmi2 function does handle it with less code, but uses goto ... so I do prefer this alternative.

**Review by modelonrobinandersson — APPROVED** [review]

**Pull request merged**